plenum/server/catchup/node_leecher_service.py

```diff
@@ -125,3 +129,46 @@ def _get_next_ledger(self, ledger_id: Optional[int]) -> Optional[int]:
             return None
 
         return ledger_ids[next_index]
+
+    def _catchup_ledger(self, ledger_id: int):
+        leecher = self._leechers[ledger_id]
+        catchup_till = self._catchup_till.get(ledger_id)
+        if catchup_till is None:
+            leecher.start(request_ledger_statuses=True)
+        else:
+            leecher.start(request_ledger_statuses=False, till=catchup_till)
+
+    def _calc_catchup_till(self, last_3pc: Optional[Tuple[int, int]]):
+        audit_ledger = self._provider.ledger(AUDIT_LEDGER_ID)
+        last_audit_txn = audit_ledger.get_last_committed_txn()
+        if last_audit_txn is None:
+            return
+
+        last_audit_txn = last_audit_txn['txn']['data']
+        view_no = last_audit_txn['viewNo']
+        pp_seq_no = last_audit_txn['ppSeqNo']
+        # view_no, pp_seq_no = last_3pc if last_3pc else (0, 0)
+        for ledger_id, final_size in last_audit_txn['ledgerSize'].items():
+            ledger = self._provider.ledger(ledger_id)
+            start_size = ledger.size
+
+            final_hash = last_audit_txn['ledgerRoot'].get(ledger_id)
+            if final_hash is None:
+                if final_size != ledger.size:
+                    raise RuntimeError('!!!')  # TODO: Write sensible message
+                final_hash = Ledger.hashToStr(ledger.tree.merkle_tree_hash(0, final_size)) if final_size > 0 else None
```

Comment:
Why can't we use `ledger.tree.root_hash` since ledger's size is equal to final_size?

Reply:
We can )